Fixed issue #12112 caused by fetch using relative paths to retrieve wasm files

chrisspen opened the pull request:
This fixes issue #12112 by patching instantiateAsync() to use an absolute URI when given just a filename.

kripken commented:
I'm not sure I understand this. Can you give a concrete example of where this would be needed? What type of URL would not work, and what the fix to it would be?

(Depending on that, maybe the existing locateFile mechanism can be used here?)

chrisspen commented:
I ran into this problem while trying to fix [this project](https://github.com/myers/chord_detector). It uses emcc to compile an index.wasm file, whose filename is later simply passed to `fetch()` as "index.wasm", which isn't a valid URI.

The odd thing is that the emcc generated code seems to request index.wasm twice. Once with the absolute URI and then later with just the filename. The first call succeeds while the later one fails, so I thought to fix checking for a bad URI and converting it.

I did see `locateFile()` and I tried using that, but it didn't seem to have any effect.

kripken commented:
If you can get a small testcase that would be good, as I think the bug may be elsewhere. We should be sending `"index.wasm"` to fetch, as we should be adding `scriptDirectory` to that first. So maybe the question is why that directory is not property set up.

chrisspen commented:
[My fork](https://github.com/chrisspen/chord_detector) of that sample project should be a reproducible test case if you're not using my patched version of emscripten.

kripken commented:
@chrisspen It's a little hard to find what to look for in that repo, sorry...

Can you create a build that shows the problem with `--profiling`? If I can inspect that I can probably see the issue in a few minutes.

chrisspen commented:
I tried adding `--profile` to the emcc line in my makefile, but that just reports:

    clang-12: warning: argument unused during compilation: '--profile' [-Wunused-command-line-argument]

However, you should be able to reproduce the error with the following steps:

    cd /tmp
    git clone git@github.com:chrisspen/chord_detector.git
    cd chord_detector/
    git submodule update --init
    npm install webworkify
    npm install serve
    source ~/git/emsdk/emsdk_env.sh # if you don't already have emcc in your path
    npm run build
    npm run serve

That should show you a URL like http://localhost:42721. Bring that up in a browser and enable the Javascript console.

You should see an error like:

    6ebc336f-c356-4930-bf44-b23aba9beebf:1845 Uncaught (in promise) TypeError: Failed to execute 'fetch' on 'WorkerGlobalScope': Failed to parse URL from index.wasm

That's because the value of `wasmBinaryFile` isn't a proper URI, causing the call to `fetch()` around line 1021 in preamble.js to fail.

If I edit preamble.js so that it does `wasmBinaryFile = locateFile(wasmBinaryFile);` right before `fetch()`, that has no effect and `wasmBinaryFile` remains `index.wasm`.

kripken commented:
I tried those steps but I get stuck at
```
$ npm run build

> chord_detector_mega@1.0.5 build ../chord_detector
> make; browserify demo.js -o bundle.js

make: *** No rule to make target 'vendor/Chord-Detector-and-Chromagram/libs/kiss_fft130/kiss_fft.c', needed by 'build/kiss_fft.o'.  Stop.
sh: browserify: command not found
```
`npm install browserify` helped but then I get
```
make: *** No rule to make target 'vendor/Chord-Detector-and-Chromagram/libs/kiss_fft130/kiss_fft.c', needed by 'build/kiss_fft.o'.  Stop.
```

However,

> I tried adding --profile`

That's a typo, it should be `--profiling`. Doing that and providing the build might be the fastest way to make progress here.

chrisspen commented:
> `npm install browserify` helped but then I get
> 
> ```
> make: *** No rule to make target 'vendor/Chord-Detector-and-Chromagram/libs/kiss_fft130/kiss_fft.c', needed by 'build/kiss_fft.o'.  Stop.
> ```

The only thing that would cause that error is if you didn't run `git submodule update --init`. Make sure `vendor/Chord-Detector-and-Chromagram` exists in your directory.

>That's a typo, it should be --profiling. Doing that and providing the build might be the fastest way to make progress here.

Ah, sorry. I'll try that.

chrisspen commented:
I've [uploaded](https://github.com/chrisspen/chord_detector/blob/master/index.js) a recompiled file generated with `--profiling` enabled.

kripken commented:
@chrisspen Please provide the full build - looks like it also needs the `.wasm` file when I try to run it (`ENOENT: no such file or directory, open '/home/azakai/Downloads/index.wasm'`).

chrisspen commented:
@kripken I did. It's right there at the link I posted. https://github.com/chrisspen/chord_detector/blob/master/index.wasm

kripken commented:
@chrisspen oh, I see... I only saw the link you posted `https://github.com/chrisspen/chord_detector/blob/master/index.js` and did not think to try replacing the `.js` with `.wasm` manually to get the other file.

I don't know how to run the build, though? I got the .js and .wasm files. I can run the .js in node but I get no error. When I create an HTML that includes the JS and run it in the browser, I get `Uncaught ReferenceError: module is not defined`.

Sorry if I'm missing something obvious here, but I think creating a build that is super-easy to test would really help in situations like this.

chrisspen commented:
Why are you creating any files? I don't think we're connecting. I've listed the exact steps to build it. You shouldn't need to create an HTML file. There's one in the project. Just follow my build instructions I listed above.

I just ran it in a clean VM and it built fine. Maybe check to see if there are any errors in the build as a result of missing dependencies? Each step should succeed. Don't proceed if something fails.

kripken commented:
I was creating a file because I was trying to download the existing build you provided. That build just has a .js and a .wasm file, so it's not clear how to run it in a browser.

I still think if you provide that build, it would be the fastest way for me to help you. I just need to run it to see the error locally, I hope.

But, I also I tried again to run the build instructions from before, in a fresh install, and once more I get
```
r$ npm run build

> chord_detector_mega@1.0.5 build /home/azakai/Downloads/temp/chord_detector
> make; browserify demo.js -o bundle.js

make: *** No rule to make target 'vendor/Chord-Detector-and-Chromagram/libs/kiss_fft130/kiss_fft.c', needed by 'build/kiss_fft.o'.  Stop.
sh: browserify: command not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! chord_detector_mega@1.0.5 build: `make; browserify demo.js -o bundle.js`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the chord_detector_mega@1.0.5 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/azakai/.npm/_logs/2020-10-13T19_49_03_434Z-debug.log
```
I made sure to run `git submodule update --init` which is in your steps - I didn't skip it.

Debugging this, I see
```
$ ls -al vendor/
total 8
drwxr-xr-x 2 azakai primarygroup 4096 Oct 13 12:48 .
drwxr-xr-x 6 azakai primarygroup 4096 Oct 13 12:48 ..
lrwxrwxrwx 1 azakai primarygroup   45 Oct 13 12:48 Chord-Detector-and-Chromagram -> /home/chris/git/Chord-Detector-and-Chromagram
```
That `/home/chris/` link looks like the problem.

Again though, please just provide the build - far simpler than getting the whole build to run on my machine.

chrisspen commented:
>That /home/chris/ link looks like the problem.

Ah, thanks. Yes that's the issue. That's why it was working for me and not you. Sorry, I'm not super familiar with git's submodule functionality. I updated the reference in .gitmodules to point to my updated repo, but apparently that's insufficient, and it still checked in the symlink I was using for local development.

I think I've fixed that now to point to the public repo on github. And yes, you still need to install browserify, which I've added to my instructions. Please try again.

>Again though, please just provide the build - far simpler than getting the whole build to run on my machine.

I'm not sure what you mean by "the build" in this context. If you mean the emscripten compiled webassembly, that's the index.wasm file. Once you `npm install ...` the three upsteam nodejs packages I listed, all you need to do to run the demo is `npm run serve`.

You don't technically need to run `npm run build`. I just thought you wanted an example of how I was creating index.wasm and generating the index.js and bundle.js files that reference it.

chrisspen commented:
I've abandoned this tool, so this bug is moot.
